From b2507020138a07d8e1e20b1a78553c8bfd3ccb86 Mon Sep 17 00:00:00 2001
From: Tommi Franttila <tommi.franttila@intel.com>
Date: Mon, 16 Feb 2015 12:55:30 +0200
Subject: [PATCH 287/307] atomisp2: Fix for address translator corruption

Input system address translation corrupts burst during
invalidate. SW workaround for this is to set burst length
manually to 128 in case of 13MPx snapshot and to 1 otherwise.

Please see also HSD:
https://hsdes.intel.com/home/default.html#article?id=1303837302

Change-Id: Icb056790192a214a960c1419bfa2e989e43bc622
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-22477
Signed-off-by: Tommi Franttila <tommi.franttila@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/42221
Reviewed-by: Vinod Govindapillai <vinod.govindapillai@intel.com>
Reviewed-by: Jukka O Kaartinen <jukka.o.kaartinen@intel.com>
Approver: Jukka O Kaartinen <jukka.o.kaartinen@intel.com>
Integrator: Wei Tao <wei.tao@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../pci/atomisp2/atomisp_driver/atomisp-regs.h      |  2 ++
 .../pci/atomisp2/atomisp_driver/atomisp_ioctl.c     | 21 +++++++++++++++++++++
 2 files changed, 23 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h
index 2ea28ce..4ce5619 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp-regs.h
@@ -201,6 +201,8 @@
 #define CSI2_PORT_C_RX_CSI_DLY_CNT_TERMEN_DLANE1		0xC4428
 #define CSI2_PORT_C_RX_CSI_DLY_CNT_SETTLE_DLANE1		0xC442C
 
+#define DMA_BURST_SIZE_REG					0xCD408
+
 #define ISP_DFS_TRY_TIMES	2
 
 #endif /* ATOMISP_REGS_H */
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
index f52f76a..1c44af0 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
@@ -1647,6 +1647,24 @@ static void atomisp_pause_buffer_event(struct atomisp_device *isp)
 	}
 }
 
+/* Input system HW workaround */
+/* Input system address translation corrupts burst during */
+/* invalidate. SW workaround for this is to set burst length */
+/* manually to 128 in case of 13MPx snapshot and to 1 otherwise. */
+static void atomisp_dma_burst_len_cfg(struct atomisp_sub_device *asd)
+{
+
+	struct v4l2_mbus_framefmt *sink;
+	sink = atomisp_subdev_get_ffmt(&asd->subdev, NULL,
+				       V4L2_SUBDEV_FORMAT_ACTIVE,
+				       ATOMISP_SUBDEV_PAD_SINK);
+
+	if (sink->width * sink->height >= 4096*3072)
+		atomisp_store_uint32(DMA_BURST_SIZE_REG, 0x7F);
+	else
+		atomisp_store_uint32(DMA_BURST_SIZE_REG, 0x00);
+}
+
 /*
  * This ioctl start the capture during streaming I/O.
  */
@@ -1685,6 +1703,9 @@ static int atomisp_streamon(struct file *file, void *fh,
 	if (pipe->capq.streaming)
 		goto out;
 
+	/* Input system HW workaround */
+	atomisp_dma_burst_len_cfg(asd);
+
 	/*
 	 * The number of streaming video nodes is based on which
 	 * binary is going to be run.
-- 
1.9.1

